app.rb

```diff
@@ -119,6 +119,30 @@ def project_path
       redirect to("/")
     end
 
+    post "/files" do
+      upload(project_path, params['file'])
+    end
+
+    post "/files/*" do
+      path = "#{project_path}/#{params['splat'].join('/')}"
+      upload(path, params['file'])
+    end
+
+    def upload(path, file)
+      directory = Directory.new(path)
+      file_path = "#{path}/#{file[:filename]}"
+
+      ::File.open(file_path, "w") do |f|
+        f.write(file[:tempfile].read)
```

Comment:
should stream from tempfile to file.  or maybe we can even mv it?  Though it's probably possibly in RAM, so stream is the safer bet.